RecoCTPPS/PixelLocal/plugins/CTPPSPixelLocalTrackProducer.cc

```diff
@@ -0,0 +1,289 @@
+#include "DataFormats/CTPPSReco/interface/CTPPSPixelLocalTrack.h"
+#include "DataFormats/DetId/interface/DetId.h"
+
+
+#include "RecoCTPPS/PixelLocal/interface/RPixDetPatternFinder.h"
+#include "RecoCTPPS/PixelLocal/interface/RPixDetTrackFinder.h"
+
+#include <string>
+#include <vector>
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include "DataFormats/CTPPSDetId/interface/CTPPSPixelDetId.h"
+
+#include "RecoCTPPS/PixelLocal/interface/RPixRoadFinder.h"
+#include "RecoCTPPS/PixelLocal/interface/RPixPlaneCombinatoryTracking.h"
+
+
+class CTPPSPixelLocalTrackProducer : public edm::stream::EDProducer<>
+{
+public:
+  explicit CTPPSPixelLocalTrackProducer(const edm::ParameterSet& parameterSet);
+ 
+  ~CTPPSPixelLocalTrackProducer() override;
+
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions & descriptions);
+
+private:
+  int verbosity_;
+  int maxHitPerPlane_;
+  int maxHitPerRomanPot_;
+  int maxTrackPerRomanPot_;
+  int maxTrackPerPattern_;
+ 
+  edm::InputTag inputTag_;
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelRecHit>> tokenCTPPSPixelRecHit_;
+  edm::ESWatcher<VeryForwardRealGeometryRecord> geometryWatcher_;
+  uint32_t numberOfPlanesPerPot_;
+  std::vector<uint32_t> listOfAllPlanes_;
+
+  std::unique_ptr<RPixDetPatternFinder> patternFinder_;
+  std::unique_ptr<RPixDetTrackFinder>   trackFinder_  ;
+
+  void run(const edm::DetSetVector<CTPPSPixelRecHit> &input, edm::DetSetVector<CTPPSPixelLocalTrack> &output);
+  
+};
+
+//------------------------------------------------------------------------------------------------//
+
+CTPPSPixelLocalTrackProducer::CTPPSPixelLocalTrackProducer(const edm::ParameterSet& parameterSet)
+{
+  inputTag_             = parameterSet.getParameter<std::string>  ("label");
+  verbosity_            = parameterSet.getUntrackedParameter<int> ("verbosity");
+  maxHitPerRomanPot_    = parameterSet.getParameter<int> ("maxHitPerRomanPot");
+  maxHitPerPlane_       = parameterSet.getParameter<int> ("maxHitPerPlane");
+  maxTrackPerRomanPot_  = parameterSet.getParameter<int> ("maxTrackPerRomanPot");
+  maxTrackPerPattern_   = parameterSet.getParameter<int> ("maxTrackPerPattern");
+  numberOfPlanesPerPot_ = parameterSet.getParameter<int> ("numberOfPlanesPerPot");
+
+  std::string patternFinderAlgorithm = parameterSet.getParameter<std::string>("patternFinderAlgorithm");
+  std::string trackFitterAlgorithm  = parameterSet.getParameter<std::string>("trackFinderAlgorithm" );
+
+  // pattern algorithm selector
+  if(patternFinderAlgorithm == "RPixRoadFinder"){
+    patternFinder_ = std::unique_ptr<RPixRoadFinder>(new RPixRoadFinder(parameterSet));
+  }
+  else{
+    throw cms::Exception("CTPPSPixelLocalTrackProducer") << "Pattern finder algorithm" << patternFinderAlgorithm << " does not exist";
+  }
+  
+  for(uint32_t i=0; i<numberOfPlanesPerPot_; ++i){
```

Comment:
reserve the size of this vector before filling it

Reply:
done